[v2] Lerna, tsdx, definitions fixes, improvements

pjamrozowicz opened the pull request:
This is a big one, sorry about that (and still work in progress, a couple of things left - those are listed at the bottom)

It includes:
- Moving the packages to lerna.
- Unifying the building and testing process by using tsdx.
- Moving tests to TypeScript. All the tests were updated and should be passing.
- Moving from TSLint to ESLint as the former will be deprecated - now there is one eslint config for all the repository.
- Improving (and probably breaking a little bit) typescript definitions - primarily for the core package. I'm not using TypeScript much, therefore, this was a challenge. I tried to see in the issues what problems are with TS but this definitely still requires work to make it perfect. I also tried to improve types that are used only internally, in plugins or the core package, to make the life of people maintaining the repo easier.
- To make the code a little bit easier to understand and extend, I tried to limit the usage of `this` binding and instead replace it with passing an instance of Rematch class to functions. Besides that I did some code cleanup. It's still not perfect, though, so I'm open to suggestions [*].
- Adding new plugins hooks - onReducer and onRootReducer.
- Changing middleware hook to createMiddleware - it is a function that returns the same thing as before but now accepts Rematch instance as the input.
- Improving the plugins - it is now possible to use both persist and immer plugins. Persist was updated to v6 and it is also now possible to use nested persisted reducers. Immer plugin has now additional config, it's possible to whitelist/blacklist reducers.
- Removing deprecated things - there are some breaking changes, therefore, I believe it's safe to remove stuff that makes the code dirty.


[*] For example, I'm not sure we need the 'exposed' property in plugins. The dispatch and effects plugins could actually not be plugins (even though it makes perfect sense) but just the integral part of the Rematch class (since they are always there anyway). This way the properties like `dispatch` or `createDispacher` are part of Rematch class. Other plugins, whenever they want to expose something, they can just assign property in Rematch class using plugin's onInit hook or add stuff to store in onStoreCreated. This would make the plugin's api easier and more typesafe I believe.

Todos:
- Update the documentation / readme / changelogs.
- Check/fix the examples.
- Update travis configuration - I'm not sure exactly how to change it - would appreciate some help with it.

I'm open to any critique and suggestions!

ShMcK commented:
Amazing, Lots of positive and heavily requested fixes/changes in here!

I'll take a look once I get free time. I'll likely be able to help with the travis config.

semoal commented:
Damn, @pjamrozowicz this looks amazing, a lot of features that we'll wanted to implement some day.

I'll take a look in detail as soon as possible.

pjamrozowicz commented:
> Damn, @pjamrozowicz this looks amazing, a lot of features that we'll wanted to implement some day.
> 
> I'll take a look in detail as soon as possible.

> Amazing, Lots of positive and heavily requested fixes/changes in here!
> 
> I'll take a look once I get free time. I'll likely be able to help with the travis config.

I'm glad you guys like the changes!
It would be great if we could push it sooner than later (if you accept it of course) so that it doesn't fall behind master branch, especially when merging other PRs, as it's a pain to keep it in sync 😉 
Nevertheless, take your time of course as it's not just a small one-liner!

semoal commented:
I'll finish reviewing it today, if @ShMcK is agree, we could create a v2.0.0 branch where locate all this code, and create a little roadmap for a stable v2.0.0 release. 
In the meantime we could release a weekly release-candidate to help people to migrate as soon as they can to the new version.
In this way, only the master branch gots hotfixes and v2.0.0 is the current pipeline of work.

ShMcK commented:
Sounds great to me. 

The travis build seems to fail because the packages do not have their dependencies installed. Previously there was an npm script `install:all` which went into each package directory and ran `npm install`. 

The project may wor locally because you've already installed the node_modules on your machine, but the CI travis starts from scratch. To  test the problem locally you could recursively delete your node modules in the folder (https://rtmccormick.com/2018/01/10/clear-node-modules-folders-recursively-mac-linux/). 

I tried installing the packages locally, but then there needs to be a way that each package installs @rematch/core 2.0 instead of the default @rematch/core 1.3 on NPM. Not sure how that maps out with lerna, but I'm getting errors as its installing from npm rather than the local `packages/core` directory.

pjamrozowicz commented:
Okay, so the first command that should be executed is actually ‘lerna bootstrap’ as it installs all the dependencies and links everything together. There’s no need to do run ‘npm install’ manually in each package.

ShMcK commented:
Great, then the minor changes I added to the travis.yml above should work.

ShMcK commented:
Strangely, the tests fail as the "redux" dependency is not found in the folders.

I've found the same issue locally. Going into "packages/core" after running `npm run bootstrap`, there is no "redux" in the node_modules.

I see in this [lerna issue](https://github.com/lerna/lerna/issues/1457) that you can run `lerna exec npm install` to install in each directory, but that also fails as there is no `@rematch/core@2.0.0` on npm.

HiMrHu commented:
sorry, I will test it at the weekend.

pjamrozowicz commented:
@ShMcK Sorry for the late response, I'm back, ready to finish delivering it :)

I run commands from the main directory in this order:
```
npm install
npm run bootstrap
```

The redux library is then located in the **root** `node_modules` directory, not in `packages/core/node_modules` because of the option `hoist: true` in `lerna.json`. It works without any issues within `/packages` as we are using `"moduleResolution": "node"` so it looks for dependencies up in the tree too.

I'm not sure why it doesn't work on travis. You could try disabling hoisting as it's just optional to speed-up installing dependencies.

Also, it looks like there will be better support for monorepos in tsdx soon (https://twitter.com/jaredpalmer/status/1204898482076766209) so I guess it's a good move to use start using tsdx 😉

ShMcK commented:
I merged a bunch of smaller PRs in the meantime, @pjamrozowicz could you resolve the conflicts and we'll get this one merged in soon?

pjamrozowicz commented:
@ShMcK Sure, I'll try to handle that over this weekend and also finish the remaining todos I have (mainly the documentation as that's crucial).

mathieutu commented on .prettierrc:
```diff
@@ -1,10 +1,10 @@
 	"singleQuote": true,
+	"tabWidth": 2,
 	"trailingComma": "es5",
+	"useTabs": true,
```
Hi, 
I was surprised to see tabs in the project. The standard is 2 spaces. Don't you want to benefit of this PR to set 2 spaces?

mathieutu commented:
Hi folks,
I'm currently working on other PRs for rematch (one fixing broken tests which in fact don't test anything, and one fixing #699), but I've just seen this one, and there will be some conflicts.

Do you plan to release this one soon, so I stop my work for now and submit it later, or do you prefer I submit it now, and we will see when this one will be merged?

My point of view is that the config function can wait if this is merged soon, but the test one should be merged before the v2, so we're sure that there isn't any unwanted breaking change.

I can also help with conflict, and refactoring on this one if you want. 
(And I've seen other non breaking couples of things I want to do on the v2, so I can do it now, or when merged)

Thanks for this cool work,
Matt'.

pjamrozowicz commented:
Hi @mathieutu!

Thanks for the suggestion about tabs, initially I left the previous prettier config but I changed it now to use spaces.

I believe there is no need to work on #699 at all as this PR fixes the conflicts between plugins and expands plugins API.

As for fixing tests, I would suggest waiting and later on just making sure the tests for rematch v2 are correct since merging changes into this branch is a real pain.

Hopefully, @ShMcK will be able to release alpha/beta version soon.
I almost got travis working, the only failing tests are the ones for compiling examples as the dependencies are not installed. I'll try to solve that in a few days and also finish the documentation.

mathieutu commented on packages/core/test/mergeConfig.test.ts:
```diff
@@ -107,45 +117,43 @@ describe('mergeConfig:', () => {
 			expect(result.redux.enhancers).toEqual([1, 2, 3, 4])
 		})
 
 		test('config redux reducers should overwrite plugin reducers', () => {
```
☝️ Actually this is "outdated" just because of spaces, but the test is still the same.

mathieutu commented:
Ok, perfect. 
Agree for #699, my first thoughts about it was same than yours.

Feel free to tell if I can help in anything (except docs, I'm afraid my English is not as good as I would like 😅)

pjamrozowicz commented:
I got back to it, hopefully, within 2-3 days I should have a finished code ready for review.

semoal commented:
> Took a bit longer than 2-3 days but I'm almost there :)
> 
> I have 2 plugins left to go over, then update the documentation website and the one in the code (I try to document each function so it's easier to work with) and remove changes for the Chinese docs as I haven't touched those but probably I messed up with a merge. Nevertheless, they will be a bit outdated for the version 2.

Probably the best option is to delete chinese documentation if you are all agree, and just maintain one source of truth. If someone doesn't understand something just google translate it, or ask via rematch slack..

pjamrozowicz commented:
@semoal I would need to remove the Chinese docs for now anyway since I'm redoing the documentation to make it cleaner and easier to follow.

Just to update anyone who might be following this thread:
- The rewrite of the library is pretty much done, it contains all the changes mentioned in the first comment. I tried to make it cleaner and easier to use.
- Even though the dispatch and effect plugins made total sense to be 'plugins', it was a nightmare to write the code with TypeScript so I moved them to be the internal part of the core. There are still some usages of `@ts-ignore` in the core but hopefully, this can be fixed easily with some small tweaks.
- The situation looks worse when it comes to plugins. I believe it's possible to make them completely type-safe but it would take me too long now so the most complicated ones (loading, select) have `@ts-nocheck`. Nevertheless, it's still more or less 'internal' thing.
- User-facing types work pretty well, especially with the changes done by @777PolarFox777 in this commit (https://github.com/rematch/rematch/commit/00d45edd8cbcc9f0d5290452c665d5d84e1816f6).
- Additionally, I'm not sure how it was working before but now the models provided by plugins, such as `updated` or `loading` are also recognized by TypeScript and show up in IDE suggestions (like here https://github.com/rematch/rematch/blob/08ca0badc42b35b09cac829b3ba2176ceb4cd849/packages/loading/test/loading-asBoolean.test.ts#L11
- All tests are in TypeScript and all are passing!

What is left:
- I need to finish writing the docs
- I need to update some in-code documentation as I was doing last-minute changes
- I want to prepare more examples: for plugins, mixing plugins together, lazily adding models, hot-reloading
- There is an issue with react-redux `connect` typings which can be seen here: https://github.com/rematch/rematch/blob/08ca0badc42b35b09cac829b3ba2176ceb4cd849/examples/count-react-ts/src/App.tsx#L47 I think it could be solved by overwriting this:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/abe7571daf1b1b864cd6d1790670f1c910d1ea34/types/react-redux/index.d.ts#L319

with this:
```
export type MapDispatchToPropsFactory<TDispatchProps, TOwnProps, TDispatch extends Dispatch> = (dispatch: TDispatch<Action>, ownProps: TOwnProps) => MapDispatchToPropsFunction<TDispatchProps, TOwnProps>;
```

But maybe it is enough to just change something in the types internally, if anyone has an idea, I would appreciate some help.

semoal commented:
> @semoal I would need to remove the Chinese docs for now anyway since I'm redoing the documentation to make it cleaner and easier to follow.
> 
> Just to update anyone who might be following this thread:
> 
> * The rewrite of the library is pretty much done, it contains all the changes mentioned in the first comment. I tried to make it cleaner and easier to use.
> * Even though the dispatch and effect plugins made total sense to be 'plugins', it was a nightmare to write the code with TypeScript so I moved them to be the internal part of the core. There are still some usages of `@ts-ignore` in the core but hopefully, this can be fixed easily with some small tweaks.
> * The situation looks worse when it comes to plugins. I believe it's possible to make them completely type-safe but it would take me too long now so the most complicated ones (loading, select) have `@ts-nocheck`. Nevertheless, it's still more or less 'internal' thing.
> * User-facing types work pretty well, especially with the changes done by @777PolarFox777 in this commit ([00d45ed](https://github.com/rematch/rematch/commit/00d45edd8cbcc9f0d5290452c665d5d84e1816f6)).
> * Additionally, I'm not sure how it was working before but now the models provided by plugins, such as `updated` or `loading` are also recognized by TypeScript and show up in IDE suggestions (like here https://github.com/rematch/rematch/blob/08ca0badc42b35b09cac829b3ba2176ceb4cd849/packages/loading/test/loading-asBoolean.test.ts#L11
> * All tests are in TypeScript and all are passing!
> 
> What is left:
> 
> * I need to finish writing the docs
> * I need to update some in-code documentation as I was doing last-minute changes
> * I want to prepare more examples: for plugins, mixing plugins together, lazily adding models, hot-reloading
> * There is an issue with react-redux `connect` typings which can be seen here: https://github.com/rematch/rematch/blob/08ca0badc42b35b09cac829b3ba2176ceb4cd849/examples/count-react-ts/src/App.tsx#L47
>    I think it could be solved by overwriting this:
>   https://github.com/DefinitelyTyped/DefinitelyTyped/blob/abe7571daf1b1b864cd6d1790670f1c910d1ea34/types/react-redux/index.d.ts#L319
> 
> with this:
> 
> ```
> export type MapDispatchToPropsFactory<TDispatchProps, TOwnProps, TDispatch extends Dispatch> = (dispatch: TDispatch<Action>, ownProps: TOwnProps) => MapDispatchToPropsFunction<TDispatchProps, TOwnProps>;
> ```
> 
> But maybe it is enough to just change something in the types internally, if anyone has an idea, I would appreciate some help.

I can help you with all the things you provided mate, if you want to open a new issue with all the pending changes with extra information, probably some first contributors will be glad to help.

I'm agree on extending react-redux typing if this generates an issue on our types.

pjamrozowicz commented:
I managed to finish writing up the docs & updating in-code comments. The last main thing to fix would be the types compatibility with react-redux (and basically any other external library for redux with types).

I did some more investigation and it looks like we don't need to augment the react-redux, but actually redux's `Dispatch` type directly. I added the following declaration to types:

```javascript
declare module 'redux' {
	export interface Dispatch<A extends Action = AnyAction> {
		<TAction extends A, TReturnType = any, TPayload = any>(
			action: TAction
		): TReturnType & {
			[key: string]: any
		}
	}
}
```

...and I don't understand why it doesn't work. It looks like the augmentation isn't applied.

How to recreate it:
- Checkout branch rematch-v2
- `npm install && npm run bootstrap && npm run build`
- Uncomment `// @ts-ignore` from **examples/count-react-ts/src/App.tsx**
- `npm run test`

Test checking correctness of examples will fail. count-react-ts example will fail to compile on this line:

```javascript
export default connect(
	mapState,
	-> mapDispatch
)(Count)
```

With the following error:
```
 Types of parameters 'dispatch' and 'dispatch' are incompatible.
              Type 'Dispatch<Action<any>>' is not assignable to type 'RematchDispatch<RootModel>'.
                Type 'Dispatch<Action<any>>' is missing the following properties from type 'ExtractRematchDispatchersFromModels<RootModel>': dolphins, sharks"
```

This is because "regular" Redux Dispatch is just a function without extra properties. However, my augmentation should make it work.

@semoal or anyone else, any idea how to fix that?

semoal commented:
> I managed to finish writing up the docs & updating in-code comments. The last main thing to fix would be the types compatibility with react-redux (and basically any other external library for redux with types).
> 
> I did some more investigation and it looks like we don't need to augment the react-redux, but actually redux's `Dispatch` type directly. I added the following declaration to types:
> 
> ```js
> declare module 'redux' {
> 	export interface Dispatch<A extends Action = AnyAction> {
> 		<TAction extends A, TReturnType = any, TPayload = any>(
> 			action: TAction
> 		): TReturnType & {
> 			[key: string]: any
> 		}
> 	}
> }
> ```
> 
> ...and I don't understand why it doesn't work. It looks like the augmentation isn't applied.
> 
> How to recreate it:
> 
> * Checkout branch rematch-v2
> * `npm install && npm run bootstrap && npm run build`
> * Uncomment `// @ts-ignore` from **examples/count-react-ts/src/App.tsx**
> * `npm run test`
> 
> Test checking correctness of examples will fail. count-react-ts example will fail to compile on this line:
> 
> ```js
> export default connect(
> 	mapState,
> 	-> mapDispatch
> )(Count)
> ```
> 
> With the following error:
> 
> ```
>  Types of parameters 'dispatch' and 'dispatch' are incompatible.
>               Type 'Dispatch<Action<any>>' is not assignable to type 'RematchDispatch<RootModel>'.
>                 Type 'Dispatch<Action<any>>' is missing the following properties from type 'ExtractRematchDispatchersFromModels<RootModel>': dolphins, sharks"
> ```
> 
> This is because "regular" Redux Dispatch is just a function without extra properties. However, my augmentation should make it work.
> 
> @semoal or anyone else, any idea how to fix that?

I tried to take a look (i'm not a typescript type expert), and i think the problem comes from `react-redux` types: 
```js
type MapDispatchToPropsParam<TDispatchProps, TOwnProps> = MapDispatchToPropsFactory<TDispatchProps, TOwnProps> | MapDispatchToProps<TDispatchProps, TOwnProps>;
    type MapDispatchToPropsFunction<TDispatchProps, TOwnProps> = (dispatch: Dispatch<Action>, ownProps: TOwnProps) => TDispatchProps;
    type MapDispatchToPropsFactory<TDispatchProps, TOwnProps> = (dispatch: Dispatch<Action>, ownProps: TOwnProps) => MapDispatchToPropsFunction<TDispatchProps, TOwnProps>;
    type MapDispatchToProps<TDispatchProps, TOwnProps> = MapDispatchToPropsFunction<TDispatchProps, TOwnProps> | TDispatchProps;
```

They use this types for fill information on connect hoc. There's an issue related to this type, https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31363
and a question on twitter with his author, https://twitter.com/apapirovski/status/1120703099822149638

Probably with this information, we have some extra clues how to fix this.
p.s: the documentation feels incredible mate, i'll take a second read later, but on a first read it's looking amazing. 
On this week, while we try to fix this issue with types, i'll prepare all the ci setup for releasing a next version for help people to migrate to, and report some errors

pjamrozowicz commented:
> I tried to take a look (i'm not a typescript type expert), and i think the problem comes from `react-redux` types:
> 
> ```js
> type MapDispatchToPropsParam<TDispatchProps, TOwnProps> = MapDispatchToPropsFactory<TDispatchProps, TOwnProps> | MapDispatchToProps<TDispatchProps, TOwnProps>;
>     type MapDispatchToPropsFunction<TDispatchProps, TOwnProps> = (dispatch: Dispatch<Action>, ownProps: TOwnProps) => TDispatchProps;
>     type MapDispatchToPropsFactory<TDispatchProps, TOwnProps> = (dispatch: Dispatch<Action>, ownProps: TOwnProps) => MapDispatchToPropsFunction<TDispatchProps, TOwnProps>;
>     type MapDispatchToProps<TDispatchProps, TOwnProps> = MapDispatchToPropsFunction<TDispatchProps, TOwnProps> | TDispatchProps;
> ```
> 
> They use this types for fill information on connect hoc. There's an issue related to this type, [DefinitelyTyped/DefinitelyTyped#31363](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31363)
> and a question on twitter with his author, https://twitter.com/apapirovski/status/1120703099822149638
> 
> Probably with this information, we have some extra clues how to fix this.
> p.s: the documentation feels incredible mate, i'll take a second read later, but on a first read it's looking amazing.
> On this week, while we try to fix this issue with types, i'll prepare all the ci setup for releasing a next version for help people to migrate to, and report some errors

Thanks!

Hmm, we could try to fix that specifically for `react-redux`, although I tried to augment it too, and still couldn't make the types be "recognizable", it was behaving as if it doesn't see any augmentation I'm doing to either `redux` or `react-redux` types.

I was also looking at `redux-thunk` which augments `redux` so that it's compatible with `react-redux` so pretty much our use case: 
https://github.com/reduxjs/redux-thunk/blob/master/src/index.d.ts#L106 but again, putting similar code in our types doesn't seem to work due to the same issue of not being recognizable... 🤔  or I am just doing something wrong

semoal commented:
> > I tried to take a look (i'm not a typescript type expert), and i think the problem comes from `react-redux` types:
> > ```js
> > type MapDispatchToPropsParam<TDispatchProps, TOwnProps> = MapDispatchToPropsFactory<TDispatchProps, TOwnProps> | MapDispatchToProps<TDispatchProps, TOwnProps>;
> >     type MapDispatchToPropsFunction<TDispatchProps, TOwnProps> = (dispatch: Dispatch<Action>, ownProps: TOwnProps) => TDispatchProps;
> >     type MapDispatchToPropsFactory<TDispatchProps, TOwnProps> = (dispatch: Dispatch<Action>, ownProps: TOwnProps) => MapDispatchToPropsFunction<TDispatchProps, TOwnProps>;
> >     type MapDispatchToProps<TDispatchProps, TOwnProps> = MapDispatchToPropsFunction<TDispatchProps, TOwnProps> | TDispatchProps;
> > ```
> > 
> > 
> > They use this types for fill information on connect hoc. There's an issue related to this type, [DefinitelyTyped/DefinitelyTyped#31363](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31363)
> > and a question on twitter with his author, https://twitter.com/apapirovski/status/1120703099822149638
> > Probably with this information, we have some extra clues how to fix this.
> > p.s: the documentation feels incredible mate, i'll take a second read later, but on a first read it's looking amazing.
> > On this week, while we try to fix this issue with types, i'll prepare all the ci setup for releasing a next version for help people to migrate to, and report some errors
> 
> Thanks!
> 
> Hmm, we could try to fix that specifically for `react-redux`, although I tried to augment it too, and still couldn't make the types be "recognizable", it was behaving as if it doesn't see any augmentation I'm doing to either `redux` or `react-redux` types.
> 
> I was also looking at `redux-thunk` which augments `redux` so that it's compatible with `react-redux` so pretty much our use case:
> https://github.com/reduxjs/redux-thunk/blob/master/src/index.d.ts#L106 but again, putting similar code in our types doesn't seem to work due to the same issue of not being recognizable... 🤔 or I am just doing something wrong

Could be the issue that types of react redux aren't inside of react-redux and they're in @types/react-redux package?
```
declare module "@types/react-redux" {}
```

pjamrozowicz commented:
> Could be the issue that types of react redux aren't inside of react-redux and they're in @types/react-redux package?
> 
> ```
> declare module "@types/react-redux" {}
> ```
> 
> nvm this doesnt work neither
> 
> edit:
> 
> if we edit this line from this:
> 
> ```
> export declare type RematchDispatch<TModels extends object = Models> = ReduxDispatch & ExtractRematchDispatchersFromModels<TModels>;
> ```
> 
> to this:
> 
> ```
> export declare type RematchDispatch<TModels extends object = Models> = ReduxDispatch & ExtractRematchDispatchersFromModels<TModels>  & Dispatch<Action<any>>;
> ```
> 
> The error disappears but we lose the autocomplete on dispatch functions... :/

Yes, I tested the `"@types/react-redux"` but without success.

I'm not sure how are you testing it because with these changes it still doesn't compile for me.

When I was working on the types, it used to be actually like this:
```javascript
export type RematchDispatch<TModels extends object = Models> = Dispatch<Action> &
	ExtractRematchDispatchersFromModels<TModels>
```
where Dispatch is original redux Dispatch and it was still failing with the error mentioned before.
But then I started to augment what everyone else sees as redux `Dispatch`, therefore I just recreated "our version" of `Dispatch` as a `ReduxDispatch` to not cause any internal type issues.

Also, we should try to figure out why the augmentation doesn't work, because our types should be alright. They work just fine for `useDispatch` hook because it doesn't strongly depend on Redux Dispatch method:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-redux/index.d.ts#L528

semoal commented:
> > Could be the issue that types of react redux aren't inside of react-redux and they're in @types/react-redux package?
> > ```
> > declare module "@types/react-redux" {}
> > ```
> > 
> > 
> > nvm this doesnt work neither
> > edit:
> > if we edit this line from this:
> > ```
> > export declare type RematchDispatch<TModels extends object = Models> = ReduxDispatch & ExtractRematchDispatchersFromModels<TModels>;
> > ```
> > 
> > 
> > to this:
> > ```
> > export declare type RematchDispatch<TModels extends object = Models> = ReduxDispatch & ExtractRematchDispatchersFromModels<TModels>  & Dispatch<Action<any>>;
> > ```
> > 
> > 
> > The error disappears but we lose the autocomplete on dispatch functions... :/
> 
> Yes, I tested the `"@types/react-redux"` but without success.
> 
> I'm not sure how are you testing it because with these changes it still doesn't compile for me.
> 
> When I was working on the types, it used to be actually like this:
> 
> ```js
> export type RematchDispatch<TModels extends object = Models> = Dispatch<Action> &
> 	ExtractRematchDispatchersFromModels<TModels>
> ```
> 
> where Dispatch is original redux Dispatch and it was still failing with the error mentioned before.
> But then I started to augment what everyone else sees as redux `Dispatch`, therefore I just recreated "our version" of `Dispatch` as a `ReduxDispatch` to not cause any internal type issues.
> 
> Also, we should try to figure out why the augmentation doesn't work, because our types should be alright. They work just fine for `useDispatch` hook because it doesn't strongly depend on Redux Dispatch type:
> https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-redux/index.d.ts#L528

Tried during 2 days, a lot of things and no one worked.
I've updated typescript from 3.8.3 to 3.9.3, and tslib to 2.0.0, and a lot of types look wrong, so probably we should update to latest from typescript and iterate over here.
https://github.com/rematch/rematch/issues/759
Same issues that people is referencing here, so we refactored the types but we have the same issues.

pjamrozowicz commented:
> Tried during 2 days, a lot of things and no one worked.
> I've updated typescript from 3.8.3 to 3.9.3, and tslib to 2.0.0, and a lot of types look wrong, so probably we should update to latest from typescript and iterate over here.
> #759
> Same issues that people is referencing here, so we refactored the types but we have the same issues.

So, I updated TS and _fixed_ the types:
- I found a "dirty" workaround for the react-redux problem with dispatch - creating the augmentation file in the user-land that lists all the models. It's not needed when just using hooks from react-redux.
- Due to the issues in #759 that came up after updating the TS version, I needed to remove 'the magic' that allowed for circular references in types. Therefore now, when declaring effects, to use the type information, user needs to use casting. I searched a lot for workarounds but I couldn't find anything working.

Both those things are described in the new docs, in the "getting started" section, when selecting TypeScript tabs.

@semoal Let me know what you think and if we can move forward

pjamrozowicz commented:
bump 😉

semoal commented:
@pjamrozowicz thanks for your amazing work mate, i'll do a hard review this weekend and we'll release candidate, i'm finishing the github action file =)

The glory is near haha

NoemiRozpara commented:
Any update here?
